### DISCUSSION TITLE
How to handle errors in ServerActions

### DISCUSSION THREAD

**Discussion by DenisBessa** [vercel/next.js#49401] [Help] [answered]
### Summary

I'm playing with ServerActions, but I'm having trouble to deal with error on ServerSide. Is there a way to return any value to the client when something goes wrong on the server?

**Comment by CogitoBurrito** [comment]
AND how to create a onSuccess callback?

**Comment by DenisBessa** [comment, marked as answer]
Thanks, guys! I've managed to return a value as a Promise on the Server Function. Just using await and I can get it on the client.

**Reply by CogitoBurrito** [reply]
When you calling the action from form action. How do you handle the error?

**Reply by DenisBessa** [reply]
I've only experimented with it using useTransition().

If you need examples, just let me know.

**Reply by CogitoBurrito** [reply]
Yes please! Would love to learn something from you.

**Reply by JoseVSeb** [reply]
wasn't the question about how to handle errors in server actions?? returning value as a promise is not error handling!?? it's the normal application flow.

**Reply by DenisBessa** [reply]
I'll try to explain what I have learned by my experience with Server Actions until now. If I'm wrong in something, please feel free to correct me.

When you call a normal API on Next.JS, it will return an 500 http response if something goes wrong in the execution. With that, you can handle the errors in your application with code like `if error 500 do stuff`.

But this won't work with Server Actions since - I think - it does not encapsulates the function logic in a way that returns a 500 response to the application if there's an error.

So, you'll have to do it by yourself. Here is an example:

```
// page.tsx

import ClientComponentThatCallsServerAction from "./ClientComponentThatCallsServerAction";

async function doStuffOnServer() {
  "use server";

  try {
    console.log("Super important stuff happening on the server");

    // uncoment to simulate an error
    // throw new Error("Something went wrong");

    return { success: true };
  } catch (error) {
    console.error("Error while doing super important stuff on the server", error);
    return { success: false, error };
  }
}

export default async function App() {
  return <ClientComponentThatCallsServerAction onClick={doStuffOnServer} />;
}
```



```
// ClientComponentThatCallsServerAction.tsx

"use client";

import { useState, useTransition } from "react";

export default function DeleteButton(props: { onClick(): Promise<{ success: boolean }> }) {
  const [isPending, startTransition] = useTransition();
  const [returnedSuccess, setReturnedSuccess] = useState<boolean | undefined>(undefined);

  const { onClick } = props;

  return (
    <div>
      <button
        type="button"
        onClick={() => {
          startTransition(async () => {
            const { success } = await onClick(); // note the await here
            if (success) {
              setReturnedSuccess(true);
            } else {
              setReturnedSuccess(false);
            }
          });
        }}
      >
        Do stuff on server!
      </button>

      {isPending && <div>Doing stuff on server...</div>}

      {returnedSuccess === true && <div>Server said success!</div>}

      {returnedSuccess === false && <div>Server said error!</div>}
    </div>
  );
}
```

**Reply by codinginflow** [reply]
Can you explain what the purpose of `useTransition` is in this example? Have you run it without `useTransition` and noticed any difference?
(I know that the docs state to use it, but they don't explain why)

**Reply by codinginflow** [reply]
Can't I just wrap the server action call in a try/catch and handle the error in catch?

**Reply by DenisBessa** [reply]
> Can't I just wrap the server action call in a try/catch and handle the error in catch?

Yes, you can do this. But keep in mind that this code, will be executed on the server. So you have to return something to the client so it "knows" an error has occurred on the server.

The difference here is that, in a normal API route, if something like an unhandled error happens on the server, it will automatically return a 500 status to the client. But this is not the behavior in server actions. It will simply won't return anything to the client. So, as you suggested, you need to wrap the function in a try/catch and return something to the client, so you can handle it on the client side.

**Reply by codinginflow** [reply]
> > Can't I just wrap the server action call in a try/catch and handle the error in catch?
> 
> Yes, you can do this. But keep in mind that this code, will be executed on the server. So you have to return something to the client so it "knows" an error has occurred on the server.
> 
> The difference here is that, in a normal API route, if something like an unhandled error happens on the server, it will automatically return a 500 status to the client. But this is not the behavior in server actions. It will simply won't return anything to the client. So, as you suggested, you need to wrap the function in a try/catch and return something to the client, so you can handle it on the client side.

But you can just throw an uncaught error in the server action and catch in on the client.

**Reply by Jared-Dahlke** [reply]
"But you can just throw an uncaught error in the server action and catch in on the client."

Nope, this will not work in production. You have to do something like 

return {error: 'my message}

you can't do `throw new Error('my message')`

i just wrote my app to do that and now have to rewrite to return the message instead of throwing

**Reply by NikitaPokryshko** [reply]
@Jared-Dahlke I have the same result. Just throwing an error from server action explicitly won't work

**Reply by JNLei** [reply]
@Jared-Dahlke  I’m having the same situation, when server action throws an error, for developing environment, the error boundary caught the error and render error.tsx correctly. However, for production, only global-error.tsx can catch the server action error. 
Do you have any more insights about this? Thanks!
